MapleServer2/Data/Static/MasteryMetadataStorage.cs

```diff
@@ -29,5 +30,16 @@ public static MasteryMetadata GetMastery(int type)
         {
             return masteries.GetValueOrDefault(type);
         }
+
+        public static int GetGradeFromXP(MasteryType type, long xp)
+        {
+            int grade = 0;
+            List<MasteryGrade> masteries = GetMastery((int) (type)).Grades;
+            while ((grade < masteries.Count) && (xp >= masteries[grade].Value))
+            {
+                grade++;
+            }
+            return grade;
+        }
```

Comment:
Linq is good. Linq is your friend.

```suggestion
        public static int GetGradeFromXP(MasteryType type, long xp)
        {
            return GetMastery((int) type).Grades.Count(x => x.Value < xp);
        }
```

Reply:
ok hail C#